### PR TITLE
feat: Added retry logic for requests if return 429 error

### PR DESCRIPTION
**Pull request by mwan-ep**
## Type

* ### Feature
  Implements a new feature

## Description

Adds retry to requests if a 429 error is returned. With each attempt the retry delay is multiplied, 

## Notes

*Any additional notes.*

### CODE REVIEW

File: src/factories/request.js [hunk 1]
```diff
@@ -50,6 +50,55 @@ const createAuthRequest = config => {
...
+  requestBody,
+  attempt = 1
+) => {
+  const maxAttempts = 4
+  const baseDelay = 1000
+  return new Promise((resolve, reject) => {
+    config.auth.fetch
+      .bind()(
+        `${config.protocol}://${config.host}/${
+          version || config.version ? `${version || config.version}/` : ''
```
[the remark this answers is not in the document]
Comment (mwan-ep) [reply src/factories/request.js:68]: I tried simplifying it but eslint doesn't allow nested ternary operators. `version ? version : config.version ? config.version : ''`

**Comment by dazadep** [comment]
Once we release an SDK version with this change, how with commerce manager pick that up ? Does that require a CM change?

**Comment by dazadep** [comment]
All the retries are identical ? is there a way for us to distinguish them on the server side? Ideally we want to be able to track the requests in datadog and see how well this change is working at least for CM. @mwan-ep we can chat if this is not clear. I can give you a quick tour of datadog

**Comment by mwan-ep** [comment]
> Once we release an SDK version with this change, how with commerce manager pick that up ? Does that require a CM change?

SDK will retry up to 4 times thus greatly increasing chances of successful call. If we wanted to handle 429 error after the SDK has failed to retry then that needs a CM change.

**Comment by dazadep** [comment]
> > Once we release an SDK version with this change, how with commerce manager pick that up ? Does that require a CM change?
> 
> SDK will retry up to 4 times thus greatly increasing chances of successful call. If we wanted to handle 429 error after the SDK has failed to retry then that needs a CM change.

let me rephrase. How will CM use the new version of js-sdk that includes this MR. Will you have to create a pull request for CM to update the version ?

**Comment by dr-ep** [comment]
@dazadep Yes, there should be a version update PR on the react-dashboard project.

File: src/factories/request.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dr-ep) [reply src/factories/request.js:68]: You can write it as:

```
    const ver = version || config.version || ''
    config.auth.fetch
      .bind()(`${config.protocol}://${config.host}/${ver}${uri}`, {
        method: method.toUpperCase(),
```

[the remark this answers is not in the document]
Comment (mwan-ep) [reply src/factories/request.js]: Updated.

**Review by dr-ep — APPROVED** [review]

**Review by dr-ep — APPROVED** [review]

File: src/config.js [hunk 2]
```diff
@@ -17,7 +17,8 @@ class Config {
       disableCart,
       reauth,
       protocol,
-      store_id
+      store_id,
+      baseDelay
```
Comment (dazadep) [inline src/config.js:21]: should this be called `retryDelay` ?

**Review by dazadep — APPROVED** [review]
